File: go/day01/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	utils "github.com/MaraudingAvenger/Advent2025/common"
)

func main() {
	lines, err := utils.ReadInputFile("inputs/day1.txt")
	if err != nil {
		panic(err)
	}

	// the dial starts at 50, and goes from 0 to 99. We want to count how many times it passes 0.
	dial := 50
	passes := 0

	const SIZE = 100

	for _, line := range strings.Fields(lines) {

		direction := line[0]
		n, err := strconv.Atoi(line[1:])
		if err != nil {
			panic(err)
		}
		distance := n

		// calculate the number of times the dial passes 0
		switch direction {
		case 'R':
			dial = (dial + distance) % SIZE
		case 'L':
			dial = (dial + SIZE - (distance % SIZE)) % SIZE
		default:
			panic(fmt.Sprintf("invalid direction: %q", direction))
		}

		if dial == 0 {
			passes++
		}
	}

	fmt.Printf("Part 1: %d\n", passes)

	dial = 50
	passes = 0
	for _, line := range strings.Fields(lines) {
		direction := line[0]
		n, err := strconv.Atoi(line[1:])
		if err != nil {
			panic(err)
		}
		distance := n

		switch direction {
		case 'R':
			for i := 0; i < distance; i++ {
				dial = (dial + 1) % SIZE
				if dial == 0 {
					passes++
				}
			}
		case 'L':
			for i := 0; i < distance; i++ {
				dial = (dial + SIZE - 1) % SIZE
				if dial == 0 {
					passes++
				}
			}
		default:
			panic(fmt.Sprintf("invalid direction: %q", direction))
		}
	}
	fmt.Printf("Part 2: %d\n", passes)
}

File: rust/day01/src/main.rs
use common::read_lines;
use common::find_input_file;

fn main() {
    let lines = read_lines(&find_input_file("inputs/day1.txt"));
    let mut dial = 50;
    let mut zero_count = 0;
    const DIAL_SIZE: i32 = 100;

    // Part 1
    for line in &lines {
        let direction = &line[0..1];
        let value: i32 = line[1..].parse().expect("Could not parse value");
        match direction {
            "L" => dial = (dial + DIAL_SIZE - value) % DIAL_SIZE,
            "R" => dial = (dial + value) % DIAL_SIZE,
            _ => panic!("Unknown direction"),
        }
        if dial == 0 {
            zero_count += 1;
        }
    }
    println!("Part 1: {}", zero_count);

    // Part 2
    dial = 50;
    zero_count = 0;
    for line in &lines {
        let direction = &line[0..1];
        let value: i32 = line[1..].parse().expect("Could not parse value");
        match direction {
            "L" => {
                for _ in 0..value {
                    dial -= 1;
                    if dial == 0 {
                        zero_count += 1;
                    }
                    if dial < 0 {
                        dial = DIAL_SIZE - 1;
                    }
                }
            }
            "R" => {
                for _ in 0..value {
                    dial += 1;
                    if dial >= DIAL_SIZE {
                        dial = 0;
                    }
                    if dial == 0 {
                        zero_count += 1;
                    }
                }
            }
            _ => panic!("Unknown direction"),
        }
    }
    println!("Part 2: {}", zero_count);
}
